src/device_manager.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: (Apache-2.0 AND BSD-3-Clause)
+
+//! System level device management.
+//!
+//! [IoManager](struct.IoManager.html) responds to manage all devices
+//! of virtual machine, register IO resources callback,
+//! unregister devices and help VM IO exit handling.
+//！
+//！VMM would take responsible for getting device resource request, ask
+//! vm_allocator to allocate the resources, ask vm_device to register the
+//! devices IO windows, and finally set resources to virtual device.
+
+use crate::resources::Resource;
+use crate::DeviceIo;
+
+use std::collections::btree_map::BTreeMap;
+use std::result;
+use std::sync::Arc;
+
+/// Error type for `IoManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The inserting device overlaps with a current device.
+    DeviceOverlap,
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System IO manager serving for all devices management and VM exit handling.
+#[derive(Default)]
+pub struct IoManager {
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<(u16, u16), Arc<dyn DeviceIo>>,
```

Comment:
Yes, let's keep it as Arc now since it can achieve real multi-thread handling as discussed  https://github.com/rust-vmm/vm-device/pull/12#discussion_r350146303

Reply:
Yes of course, `Arc` is fine, and I agree this allows multi-threading, but at some point, we might need two different thread to modify the instance. That's where you need `Mutex` to be added to the picture.

Reply:
The idea is to move lock into the `dyn DeviceIo` object, and let the device driver make decision how to protect itself from concurrent access.